Bluetooth: host: Initial support for BIS

Thalley opened the pull request:
Added initial support for BIS (broadcast ISO stream), which
adds support for creating BISes as both broadcaster and receiver,
as well as managing and creating BIGs.

Extends PA sync to handle BIGInfo adveritising reports.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: host: Initial support for BIS

Added initial support for BIS (broadcast ISO stream), which
adds support for creating BISes as both broadcaster and receiver,
as well as managing and creating BIGs.

Extends PA sync to handle BIGInfo adveritising reports.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -1093,3 +1102,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	/* Transform from bitfield to array */
+	for (int i = 0; i < 0x1F; i++) {
+		if (param->bis_bitfield & BIT(i)) {
+			if (bit_idx == big->num_bis) {
```
Should this "==" be a ">="?

asbjornsabo reviewed (commented):
Seems sound to me.  I have some comments, though.

Thalley commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -1093,3 +1102,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+
+		if (!bis) {
+			BT_DBG("BIS was NULL");
+			return -EINVAL;
```
Your update is correct :)

Thalley committed:
Bluetooth: shell: Adds BT_INTERVAL_TO_MS macro

Added a simple macro for the BT shell that transforms the
interval unit (N * 1.25ms) to ms.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -474,10 +472,30 @@ static void per_adv_sync_recv_cb(
+		    "framing 0x%02x, %sencrypted",
+		    bt_le_per_adv_sync_get_index(sync), le_addr, biginfo->sid,
+		    biginfo->num_bis, biginfo->nse, biginfo->iso_interval,
+		    biginfo->iso_interval * 5 / 4, biginfo->bn, biginfo->pto,
```
Transforms the interval from N * 1.25ms to ms. Added a macro for it, and updated the other places in this file with the same calculation.

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -229,14 +229,211 @@ static int cmd_disconnect(const struct shell *shell, size_t argc,
+{
+	int err;
+	struct bt_iso_create_big_param_t param;
+	struct bt_le_ext_adv *adv = adv_sets[selected_adv];
```
Global variables, used for all ext_adv commands

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -229,14 +229,211 @@ static int cmd_disconnect(const struct shell *shell, size_t argc,
 	SHELL_CMD_ARG(connect, NULL, "Connect ISO Channel", cmd_connect, 1, 0),
 	SHELL_CMD_ARG(listen, NULL, "[security level]", cmd_listen, 1, 1),
-	SHELL_CMD_ARG(send, NULL, "Send to ISO Channel", cmd_send, 1, 0),
+	SHELL_CMD_ARG(send, NULL, "Send to ISO Channel", cmd_send, 1, 1),
```
Moved to separate commit.

Vudentz commented on include/bluetooth/iso.h:
```diff
@@ -79,12 +79,13 @@ struct bt_iso_chan_qos {
 	 *  Possible values: BT_ISO_CHAN_QOS_IN, BT_ISO_CHAN_QOS_OUT or
-	 *  BT_ISO_CHAN_QOS_INOUT.
+	 *  BT_ISO_CHAN_QOS_INOUT. Shall be BT_ISO_CHAN_QOS_OUT for broadcast
+	 *  transmitting, and BT_ISO_CHAN_QOS_IN for broadcast receiver.
```
Perhaps it would be a good idea to have separate direction for Broadcast e.g.: BT_ISO_CHAN_QOS_RECEIVER so it clearly indicates to users what kind of channel transport it is using.

Thalley replied:
Interesting idea. I'm a bit concerned that having both a `IN` and a `RECEIVER`, as well as an `OUT` and a `BROADCASTER`, might feel like duplication for some. 

I think a better solution might be to do as you suggest below with a broadcast and a unicast struct, and simply leave out the `dir` for the broadcaster. I guess we could also simply ignore the `dir` QOS value for broadcasting, as it is very clear what the ISO will do, as there are separate APIs for broadcaster and receiver.

Vudentz commented on include/bluetooth/iso.h:
```diff
@@ -79,12 +79,13 @@ struct bt_iso_chan_qos {
 	/** Channel interval */
 	uint32_t			interval;
-	/** Channel SCA */
+	/** Channel SCA - Only for CIS */
```
Perhaps we should split this into it own struct and then have a union for broadcast and unicast specific parameters.

Thalley replied:
Not a bad idea. 

Do you mean something like:

```
struct bt_iso_chan_qos {
    union {
        struct {
            <CIS fields>
        } cis;

        struct {
            <BIS fields>
        } bis;
    };
};
```

Vudentz commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -99,10 +99,23 @@ struct bt_conn_sco {
+	};
+
+	/** If true, this is a ISO for a BIS, else it is a ISO for a CIS */
+	bool is_bis;
```
We could use the presence of an ACL to indicate if this is a broadcast or a unicast ISO handle.

Thalley replied:
Took me a while to figure out why I didn't just do that. It looks like we can be in a case where acl is NULL, but while it is still a CIS (e.g. when creating CIS'es fails), and this value is currently only being used in the cleanup. So if binding a conn and a iso channel fails, the `acl` pointer cannot be used.

Vudentz commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -34,7 +34,19 @@ config BT_AUDIO_UNICAST
 	  Isochronous channels.
 
-if BT_AUDIO_UNICAST
+config BT_AUDIO_BROADCAST
```
While we do depend on bt_conn it should be possible to decouple from BT_CONN, the problem is that we will probably need to extract part of conn.c in order to make the packet fragmentation, etc, work with broadcast alone, or perhaps we do the other way around and decouple the ACL handling from conn.c as with the introduction of BIS it is not possible to have handles/bt_conn that don't map to ACL connections, that said it perhaps would be better to rename BT_CONN to BT_ACL and have BT_CONN as a hidden option that gets enabled when it BT_ACL or BT_ISO is enabled.

Vudentz commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4828,6 +4828,45 @@ static void le_past_received(struct net_buf *buf)
+static void hci_le_biginfo_adv_report(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_biginfo_advertising_report *evt =
+		(struct bt_hci_evt_le_biginfo_advertising_report *)buf->data;
```
This looks duplicated since you will be doing net_buf_pull_mem and overwriting the original pointer, Id leave just the later since it also adjust the buffer position.

Thalley replied:
Nice catch :)

Vudentz commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -237,6 +422,18 @@ SHELL_STATIC_SUBCMD_SET_CREATE(iso_cmds,
 	SHELL_CMD_ARG(disconnect, NULL, "Disconnect ISO Channel",
 		      cmd_disconnect, 1, 0),
+#if defined(CONFIG_BT_ISO_BROADCAST)
+	SHELL_CMD_ARG(big-create, NULL,
```
I usually avoid using the same prefix for multiple commands to make tab completion more efficient, either move to be a suffix e.g. create-big or add a subcmd big.

Thalley replied:
Sure, we can do that.

Thalley commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -256,32 +259,40 @@ void bt_iso_cleanup(struct bt_conn *conn)
 struct bt_conn *iso_new(void)
 {
-	return bt_conn_new(iso_conns, ARRAY_SIZE(iso_conns));
+	struct bt_conn *iso = bt_conn_new(iso_conns, ARRAY_SIZE(iso_conns));
```
In the current implementation (PR) of ISO in the Zephyr controller, we have separate channels for BIS and CIS, but in the host we have a single pool (`iso_conns`). I guess it would make sense to align these; what makes most sense? A shared pool, or separate pools? 

If we want to support switching from CIS to BIS, or vice versa, I suppose a shared pool makes most sense?

Vudentz commented on include/bluetooth/iso.h:
```diff
@@ -245,6 +342,41 @@ int bt_iso_chan_disconnect(struct bt_iso_chan *chan);
+ */
+int bt_iso_create_big(struct bt_le_ext_adv *padv,
+		      struct bt_iso_create_big_param_t *param);
+
```
I would attemp to map the parameters to bt_iso_channel as well, so we can use function like bt_iso_disconnect to terminate, etc, also perhaps we should name this bt_iso_broadcast(channel, padv) then the channel would contain qos with the related field to create a the big. If this all can be arranged it would also be possible to use bt_iso_chan_send to transmit over broadcast.

Thalley replied:
`bt_iso_chan_send` is already used to send over broadcast. 

Regarding `bt_iso_terminate` that should also be possible to implement. The reason why I haven't done that, is that I don't see a use case where you'd want to disconnect a single BIS in a BIG (actually I don't understand the usecase for disconnecting a single CIS in a CIG). Do you have an example of where you'd might want to do that? 

I'm not sure what other parameters you are referring to here.

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5021,6 +5059,17 @@ static const struct event_handler meta_events[] = {
+		      sizeof(struct bt_hci_evt_le_big_sync_lost)),
+	EVENT_HANDLER(BT_HCI_EVT_LE_BIGINFO_ADV_REPORT,
+		      hci_le_biginfo_adv_report,
+		      sizeof(struct bt_hci_evt_le_biginfo_advertising_report)),
```
```suggestion
		      sizeof(struct bt_hci_evt_le_biginfo_adv_report)),
```

cvinayak commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -1090,3 +1099,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+static void cleanup_big(struct bt_iso_big *big)
+{
+	for (int i = 0; i < big->num_bis; i++) {
+		bt_iso_cleanup(big->bis[i]->conn);
```
Why is connection cleanup?

cvinayak commented on include/bluetooth/iso.h:
```diff
@@ -118,6 +119,102 @@ struct bt_iso_chan_path {
 
+struct bt_iso_big {
+	/** Array of ISO channels to setup as BIS (the BIG). */
+	struct bt_iso_chan **bis;
```
You should not associate a ACL connection with BIG ISO channel

asbjornsabo commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	} else {
+		bigs[big->handle] = NULL;
+		cleanup_big(big);
+	}
```
Missing the big picture: What is done if there is an error terminating?

Thalley replied:
For BIGs this actually cannot fail (only internal host states are updated, and all functions are even `void`).

asbjornsabo replied:
I was thinking about the error returned from the hci commands.

Thalley replied:
There are no HCI commands for cleanup_big :) All cleanup is done on the host side only. `cleanup_big` is more often than not, used to actually handle host cleanup if any HCI commands fail.

asbjornsabo commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_BIG_CREATE_SYNC, buf, NULL);
+	if (err) {
```
Superfluous lines - err is returned in any case

Thalley replied:
Good catch

asbjornsabo commented on subsys/bluetooth/host/iso_internal.h:
```diff
@@ -11,6 +11,7 @@
 #include <bluetooth/iso.h>
 
 #define BT_ISO_DATA_PATH_DISABLED			0xFF
+#define BT_ISO_INVALID_BIG_HANDLE			0xFF
```
Are these supposed to have the same value?

Thalley replied:
They are just internal values, that represent some invalid value. It's okay that they are the same :)

asbjornsabo commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -39,6 +39,9 @@
 
+/* Multiply bt 1.25 to get MS */
+#define BT_INTERVAL_TO_MS(interval) ((interval) * 5 / 4)
+
```
As discussed in another context, this will only give rounded values for the interval.
Still not completely happy about that ...

Is it possible to ask whoever did this change (convert to integer) all-over what the reason is?

(Not that important, though, it is only in the shell.)

Thalley replied:
@jhedberg You made the initial change to this it seems in https://github.com/zephyrproject-rtos/zephyr/commit/cbb7ef55600220a99bc7cac69d66c5acf1fd05f3. Any comments?

jhedberg replied:
@Thalley doesnt the commit message explain the reasoning?

Thalley replied:
I think so. @asbjornsabo do you agree?

jhedberg replied:
Alternatively, instead of trying to do the conversion in code, the output could be changed to something like:
```
shell_print(ctx_shell, "Interval: 0x%04x x 1.25ms", info.le.interval);
```
I.e. let the user do the calculation if they want to.

asbjornsabo reviewed (commented):
Not fully understanding this yet, but it seems sane to me.
Some minor comments.

Thalley commented on include/bluetooth/iso.h:
```diff
@@ -245,6 +342,41 @@ int bt_iso_chan_disconnect(struct bt_iso_chan *chan);
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_iso_create_big(struct bt_le_ext_adv *padv,
```
I agree that they at least should be similar (names were taken after HCI commands to begin with). Regarding the order of Zephyr, there is not "one way". See https://github.com/zephyrproject-rtos/zephyr/pull/30892/files#r552095727 for a bit more discussion on the subject.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4745,6 +4745,47 @@ static void le_past_received(struct net_buf *buf)
 #endif /* CONFIG_BT_CONN */
+
+#if defined(CONFIG_BT_ISO_BROADCAST)
+static void hci_le_biginfo_adv_report(struct net_buf *buf)
```
This is a static function; only used by a event handler below.

Thalley commented on include/bluetooth/iso.h:
```diff
@@ -118,6 +119,102 @@ struct bt_iso_chan_path {
+	/** Bitfield of the BISes to sync to */
+	uint32_t bis_bitfield;
+
+	/** @brief Maximum subevetns
```
Fixed

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5586,6 +5638,20 @@ static int le_set_event_mask(void)
 		}
 	}
 
+	/* Enabled BIS events for broadcaster and/or receiver */
```
Fixed

cvinayak commented on include/bluetooth/iso.h:
```diff
@@ -118,6 +119,102 @@ struct bt_iso_chan_path {
+	uint16_t max_pdu;
+
+	/** The interval, in microseconds, of periodic SDUs. */
+	uint16_t sdu_interval;
```
```suggestion
	uint32_t sdu_interval;
```

asbjornsabo reviewed (approved):
LGTM, as far as I understand it.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,424 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	int err;
+	bool broadcaster;
+
+	if (!big->handle || !big->num_bis || !big->bis) {
```
0 is a valid BIG handle

Thalley replied:
Right. I actually already had a #define for invalid handles that I guess wasn't used here.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,424 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	if (evt->big_handle != big->handle) {
+		err = -EIO;
+	} else {
+		big_disconnect(big);
```
Have to wait for HCI LE BIG Terminate Complete event.

Thalley replied:
Interesting that I didn't implement that. Reimplementing it feels like I did that already, but for some reason must have removed that code again. 

Anyways, I readded the handling of the `BT_HCI_EVT_LE_BIG_TERMINATE` event such that we, as the broadcaster, don't clean up the BIG in the host until we receive that event.

cvinayak reviewed (dismissed):
To void too much rework due to maintaining/resolve possible merge conflict on the existing implementation. This PR is approved as-is with the clause that this is experimental work in progress feature.

Thalley commented:
@cvinayak  
> cleanup_big is segmentation faulting, suspecting its is access ACL data members?

That shouldn't happen if the `bt_conn` represents a BIG. Can you give any more information for this?

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,442 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		return err;
+	}
+
+	evt = (struct bt_hci_evt_le_big_complete *)rsp->data;
+	if (evt->status ||
+	    evt->big_handle != big->handle ||
+	    evt->num_bis != big->num_bis) {
+		err = -EIO;
+	} else {
+		for (int i = 0; i < big->num_bis; i++) {
+			big->bis[i]->conn->handle = evt->handle[i];
+			bt_iso_chan_set_state(big->bis[i], BT_ISO_CONNECTED);
+		}
+
+		bigs[big->handle] = big;
+	}
```
Here `rsp` is command response, not LE Create BIG Complete event.
LE Create BIG has a command status followed by an asynchronous LE Create BIG Complete event.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,454 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	struct bt_hci_cp_le_terminate_big *req;
+	struct bt_hci_evt_le_big_terminate *evt;
+	struct net_buf *buf;
+	struct net_buf *rsp;
```
```suggestion
```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,454 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	req->big_handle = big->handle;
+	req->reason = BT_HCI_ERR_REMOTE_USER_TERM_CONN;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_TERMINATE_BIG, buf, &rsp);
```
```suggestion
	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_TERMINATE_BIG, buf, NULL);
```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,454 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		return err;
+	}
+
+	evt = (struct bt_hci_evt_le_big_terminate *)rsp->data;
+	if (evt->big_handle != big->handle) {
+		err = -EIO;
+	}
+
```
```suggestion

```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,454 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		memset(req->bcode, 0, sizeof(req->bcode));
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_CREATE_BIG, buf, &rsp);
```
```suggestion
	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_CREATE_BIG, buf, NULL);
```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,454 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	struct bt_hci_cp_le_create_big *req;
+	struct bt_iso_big *big = param->big;
+	struct net_buf *buf;
+	struct net_buf *rsp;
```
```suggestion
```

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,454 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+
+void hci_le_big_complete(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_big_complete *evt = (struct bt_hci_evt_le_big_complete *)buf->data;
```
I've seen this done a lot, but I don't understand the point of casting it to void, rather than the actual type you want it to be. Can you elaborate?

cvinayak replied:
> The conversion from void pointer to any other pointer type is guaranteed by the C programming language.

This is as stated in the Linux coding style documentation.

This discussion comes up a lot, you can take an example from `memcpy`, its `dest` and `src` parameters are `void *`. In essence, as long as we intend to typecast, irrespective of word alignments etc., the structure members will be accessed using the assigned actual pointer type variable. As long as C programming language will do the right thing, overloading data using `void *` is commonly followed.

Thalley replied:
I agree that it's common, but I still don't quite understand why, or if, it's preferred. 

To me it makes more sense to have lines like 
```
<type> var = (<type>)value;
```
than 
```
<type> var = (void *)value;
```
as the former is more precise in the casting. 

Nonetheless I've changed it to void casts to be consistent.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,442 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		return err;
+	}
+
+	*out_big = big;
```
```suggestion
	atomic_set(&big->initialized, true);
	
	*out_big = big;
```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,442 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	/* Transform from bitfield to array */
+	for (int i = 0; i < 0x1F; i++) {
+		if (param->bis_bitfield & BIT(i)) {
+			if (bit_idx == big->num_bis) {
```
```suggestion
			if ((bit_idx + 1) == big->num_bis) {
```

Thalley replied:
I am not sure I understand this suggestion. 

If `big->num_bis == 1`, then this will then always fall as `bit_idx + 1 == big->num_bis`.

cvinayak replied:
my fault, this was a hack to workaround the uninitialized `big->num_bis` during my testing.

Could you fix this, and the missing disconnected callback on sync loss.

Thalley replied:
I think the missing callback was due to the missing reference to the ISO channels (as they define the callbacks). Updated it now by adding reference to the ISO channels to `bt_iso_big_sync_param_t`. Please test.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,442 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		return err;
+	}
+
+	err = hci_le_big_create_sync(sync, *out_big, param);
```
```suggestion
	err = hci_le_big_create_sync(sync, big, param);
```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1096,3 +1105,442 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		bt_iso_chan_set_state(big->bis[i], BT_ISO_CONNECT);
+	}
+
+	*out_big = big;
```
```suggestion
	atomic_set(&big->initialized, true);
	
	*out_big = big;
```
